File: Pharmaceutical_Company_Example/Behavioral/Python/State.py
from abc import ABC, abstractmethod

# State interface
class DrugState(ABC):
    @abstractmethod
    def proceed_to_next(self, drug_application) -> None:
        pass

    @abstractmethod
    def get_state_name(self) -> str:
        pass

# Submitted State
class SubmittedState(DrugState):
    def proceed_to_next(self, drug_application) -> None:
        print("Proceeding from Submitted to Clinical Trial...")
        drug_application.set_state(ClinicalTrialState())

    def get_state_name(self) -> str:
        return "Submitted"

# Clinical Trial State
class ClinicalTrialState(DrugState):
    def proceed_to_next(self, drug_application) -> None:
        print("Proceeding from Clinical Trial to Under Review...")
        drug_application.set_state(UnderReviewState())

    def get_state_name(self) -> str:
        return "Clinical Trial"

# Under Review State
class UnderReviewState(DrugState):
    def proceed_to_next(self, drug_application) -> None:
        print("Proceeding from Under Review to Approved...")
        drug_application.set_state(ApprovedState())

    def get_state_name(self) -> str:
        return "Under Review"

# Approved State
class ApprovedState(DrugState):
    def proceed_to_next(self, drug_application) -> None:
        print("The drug has already been approved. No further states.")

    def get_state_name(self) -> str:
        return "Approved"

# Context class
class DrugApplication:
    def __init__(self, state: DrugState) -> None:
        self._state = state

    def set_state(self, new_state: DrugState) -> None:
        self._state = new_state
        print(f"DrugApplication: State changed to: {self._state.get_state_name()}")

    def proceed(self) -> None:
        self._state.proceed_to_next(self)

    def get_current_state(self) -> str:
        return self._state.get_state_name()

# Initial drug application in "Submitted" state
drug_application = DrugApplication(SubmittedState())

drug_application.proceed()  # Moves to Clinical Trial
drug_application.proceed()  # Moves to Under Review
drug_application.proceed()  # Moves to Approved
drug_application.proceed()  # Already Approved, no further state

File: Pharmaceutical_Company_Example/Behavioral/C++/State.cpp
#include <iostream>
#include <string>

// State interface
class DrugState {
public:
  virtual void proceed_to_next(class DrugApplication& drug_application) = 0;
  virtual std::string get_state_name() const = 0;
  virtual ~DrugState() = default;
};

// Forward declarations of concrete state classes
class SubmittedState;
class ClinicalTrialState;
class UnderReviewState;
class ApprovedState;

// Context class: DrugApplication
class DrugApplication {
private:
  DrugState* state;
  
public:
  DrugApplication(DrugState* initial_state) : state(initial_state) {}
  
  void set_state(DrugState* new_state) {
    state = new_state;
    std::cout << "DrugApplication: State changed to: " << state->get_state_name() << std::endl;
  }
  
  void proceed() {
    state->proceed_to_next(*this);
  }
  
  std::string get_current_state() const {
    return state->get_state_name();
  }
};

// Submitted State
class SubmittedState : public DrugState {
public:
  void proceed_to_next(DrugApplication& drug_application) override {
    std::cout << "Proceeding from Submitted to Clinical Trial..." << std::endl;
    drug_application.set_state(new ClinicalTrialState());
  }
  
  std::string get_state_name() const override {
    return "Submitted";
  }
};

// Clinical Trial State
class ClinicalTrialState : public DrugState {
public:
  void proceed_to_next(DrugApplication& drug_application) override {
    std::cout << "Proceeding from Clinical Trial to Under Review..." << std::endl;
    drug_application.set_state(new UnderReviewState());
  }
  
  std::string get_state_name() const override {
    return "Clinical Trial";
  }
};

// Under Review State
class UnderReviewState : public DrugState {
public:
  void proceed_to_next(DrugApplication& drug_application) override {
    std::cout << "Proceeding from Under Review to Approved..." << std::endl;
    drug_application.set_state(new ApprovedState());
  }
  
  std::string get_state_name() const override {
    return "Under Review";
  }
};

// Approved State
class ApprovedState : public DrugState {
public:
  void proceed_to_next(DrugApplication& drug_application) override {
    std::cout << "The drug has already been approved. No further states." << std::endl;
  }
  
  std::string get_state_name() const override {
    return "Approved";
  }
};

int main() {
  // Initial drug application in "Submitted" state
  DrugApplication drug_application(new SubmittedState());
  
  drug_application.proceed();  // Moves to Clinical Trial
  drug_application.proceed();  // Moves to Under Review
  drug_application.proceed();  // Moves to Approved
  drug_application.proceed();  // Already Approved, no further state
  
  return 0;
}
